frontends/common/constantFolding.cpp

```diff
@@ -761,7 +761,7 @@ const IR::Node* DoConstantFolding::shift(const IR::Operation_Binary* e) {
 
     auto tb = left->type->to<IR::Type_Bits>();
     if (tb != nullptr) {
-        if (((unsigned)tb->width_bits() < shift) && warnings)
+        if (((unsigned)tb->width_bits() <= shift) && warnings)
             ::warning(ErrorType::WARN_OVERFLOW, "%1%: Shifting %2%-bit value with %3%", e,
```

Comment:
From user's perspective, I'm not sure if printing the value here is useful. For example, the following code snippet
```P4
bit<5> f(in bit<16> a)
{
  return (5w1) << 2048;
}
```
would have warning:
> shiftoverflow.p4(3): [--Wwarn=mismatch] warning: 5w32317006071311007300714876688669951960444102669715484032130345427524655138867890893197201411522913463688717960921898019494119559150490921095088152386448283120630877367300996091750197750389652106796057638384067568276792218642619756161838094338476170470581645852036305042887575891541065808607552399123930385521914333389668342420684974786564569494856176035326322058077805659331026192708460314150258592864177116725943603718461857357598351152301645904403697613233287231227125684710820209725157101726931323469678542580656697935045997268352998638215525166389437335543602135433229604645318478604952148193555853611059596230656: value does not fit in 5 bits

Perhaps we could use more a succinct warning like `Left shift count >= width of type`?

Reply:
The error message you are quoting comes from a different place in the code, which checks for overflow in assignments (or in general, casting a large value to a narrower bitwidth).
If you think that the message can be improved you should probably file a new issue.